internal/core/server/kernel.go

```diff
@@ -28,6 +30,45 @@ type Kernel struct {
 	TaskTracker *task.TaskTracker
 }
 
+func (kern *Kernel) DescribeTemplates(ctx context.Context, input *api.DescribeTemplatesInput) (*api.DescribeTemplatesOutput, error) {
+	return &api.DescribeTemplatesOutput{Templates: template.GetTemplateDescriptions()}, nil
+}
+
+func (kern *Kernel) CreateProject(ctx context.Context, input *api.CreateProjectInput) (*api.CreateProjectOutput, error) {
+	projectDir := input.Root
+	if !filepath.IsAbs(projectDir) {
+		return &api.CreateProjectOutput{}, errors.New("path must be absolute")
+	}
+
+	var templateDir string
+	if input.TemplateUrl != nil {
+		var err error
+		templateDir, err = template.GetTemplateFiles(ctx, *input.TemplateUrl)
+		if err != nil {
+			return &api.CreateProjectOutput{}, fmt.Errorf("getting template files: %w", err)
+		}
+	}
+
+	err := os.Mkdir(projectDir, 0750)
+	if err != nil {
+		return &api.CreateProjectOutput{}, fmt.Errorf("making project dir: %w", err)
+	}
+
+	if templateDir != "" {
+		copy.Copy(templateDir, projectDir, copy.Options{
```

Comment:
Why do we need to uncompress in to a temporary directory and then copy in to the project directory? Why not directly uncompress in to the project directory?

Reply:
I changed it from a copy to a move but the same question still applies. I believe it's conventional for a tarball of a directory to contain the directory itself at the top level. That means that we need to unwrap the containing directory when extracting the tarball. We could do this in the extraction logic but it seemed simpler and clearer to leave that logic untouched and unwrap it using a temporary directory. Happy to change if you think it would be better.

Reply:
i've seen tarballs both way, but yeah, it's usually nice to have one top-level directory so that you don't wind pu with a ton of files mixed in with your other stuff when you untar